backend/operations_scanner.go

```diff
@@ -0,0 +1,346 @@
+package main
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log/slog"
+	"net/http"
+	"os"
+	"time"
+
+	ocmsdk "github.com/openshift-online/ocm-sdk-go"
+	cmv1 "github.com/openshift-online/ocm-sdk-go/clustersmgmt/v1"
+	ocmerrors "github.com/openshift-online/ocm-sdk-go/errors"
+
+	"github.com/Azure/ARO-HCP/internal/api/arm"
+	"github.com/Azure/ARO-HCP/internal/database"
+	"github.com/Azure/ARO-HCP/internal/ocm"
+)
+
+const (
+	defaultCosmosOperationsPollInterval = 30 * time.Second
+	defaultClusterServicePollInterval   = 10 * time.Second
+)
+
+type OperationsScanner struct {
+	dbClient         database.DBClient
+	lockClient       *database.LockClient
+	clusterService   ocm.ClusterServiceClient
+	activeOperations []*database.OperationDocument
+	done             chan struct{}
+}
+
+func NewOperationsScanner(dbClient database.DBClient, ocmConnection *ocmsdk.Connection) *OperationsScanner {
+	return &OperationsScanner{
+		dbClient:         dbClient,
+		lockClient:       dbClient.GetLockClient(),
+		clusterService:   ocm.ClusterServiceClient{Conn: ocmConnection},
+		activeOperations: make([]*database.OperationDocument, 0),
+		done:             make(chan struct{}),
+	}
+}
+
+func getInterval(envName string, defaultVal time.Duration, logger *slog.Logger) time.Duration {
+	if intervalString, ok := os.LookupEnv(envName); ok {
+		interval, err := time.ParseDuration(intervalString)
+		if err == nil {
+			return interval
+		} else {
+			logger.Warn(fmt.Sprintf("Cannot use %s: %s", envName, err.Error()))
+		}
+	}
+	return defaultVal
+}
+
+func (s *OperationsScanner) Run(ctx context.Context, logger *slog.Logger, stop <-chan struct{}) {
```

Comment:
What does it mean to take a context _and_ a stop channel? It looks like you could use `ctx.Done()` instead of needing `stop`.

Reply:
I wasn't aware of `signal.NotifyContext` and I can see the use cases for it, but I think I'm going to decline using it here.  The intent of the `stop` channel is to stop the timers upon receiving a terminate signal but let the last scan run to completion.  Cancelling the context on a terminate signal would abort the scanner immediately and potentially leave Cosmos data across multiple containers in an inconsistent state (though I realize this is still possible if the pod is forcefully killed).

Reply:
If you truly have two individual and distinct types of termination (why? is that really necessary? as you note, the application needs to be robust to this in any case, and if possible transactions etc need to be used), you should still use `context.Context` - that's why it exists. Passing through a `stop` channel is a much older and less powerful mechanism to achieve it. Mixing both approaches is pretty complicated.

Reply:
As I understand, Cosmos transactions are limited to a single partition in a single container.  In our case we're juggling two containers with item state that needs to stay consistent.  (This may be a sign we need to rethink our database design, but that won't be addressed here.)